control/freqplot.py

```diff
@@ -1039,14 +1040,173 @@ def gangof4_plot(P, C, omega=None, **kwargs):
 
     plt.tight_layout()
 
+#
+# Singular value plot
+#
+
+
+def singular_values_plot(syslist, omega=None,
+                         plot=True, omega_limits=None, omega_num=None,
+                         *args, **kwargs):
+    """Singular value plot for a system
+
+    Plots a Singular Value plot for the system over a (optional) frequency range.
+
+    Parameters
+    ----------
+    syslist : linsys
+        List of linear systems (single system is OK)
+    omega : array_like
+        List of frequencies in rad/sec to be used for frequency response
+    plot : bool
+        If True (default), plot magnitude and phase
+    omega_limits : array_like of two values
+        Limits of the to generate frequency vector.
+        If Hz=True the limits are in Hz otherwise in rad/s.
+    omega_num : int
+        Number of samples to plot.  Defaults to
+        config.defaults['freqplot.number_of_samples'].
+
+    Returns
+    -------
+    sigma : ndarray (or list of ndarray if len(syslist) > 1))
+        singular values
+    omega : ndarray (or list of ndarray if len(syslist) > 1))
+        frequency in rad/sec
+
+    Other Parameters
+    ----------------
+    grid : bool
+        If True, plot grid lines on gain and phase plots.  Default is set by
+        `config.defaults['bode.grid']`.
+
+    Examples
+    --------
+    >>> den = [75, 1]
+    >>> sys = ct.tf([[[87.8], [-86.4]], [[108.2], [-109.6]]], [[den, den], [den, den]])
+    >>> omega = np.logspace(-4, 1, 1000)
+    >>> sigma, omega = singular_values_plot(sys)
```

Comment:
I have to admit I am not too familiar with doctest. I thought about something like this:

    >>> den = [75, 1]
    >>> sys = ct.tf([[[87.8], [-86.4]], [[108.2], [-109.6]]], [[den, den], [den, den]])
    >>> omega = np.logspace(-4, 1, 1000)
    >>> sigma, omega = singular_values_plot(sys)
    >>> singular_values_plot(sys, 0.0, plot=False)
    (array([[197.20868123],
        [  1.39141948]]),
     array([0.]))

Since it is a plotting function, I'd like to keep as first an example that produces a figure and does have text output 
(would be too long...) and as second example one that produces a small output (evaluation at one point) that can be checked. Makes sense?

Reply:
```
[ben@skylab:…python-control]% pytest --doctest-modules -k singular_values_plot control/freqplot.py                            [1]
====================================================== test session starts =======================================================
platform linux -- Python 3.8.8, pytest-6.2.2, py-1.10.0, pluggy-0.13.1
Matplotlib: 3.4.1
Freetype: 2.6.1
rootdir: /home/ben/src/python-control, configfile: setup.cfg
plugins: timeout-1.4.2, mpl-0.12, xvfb-2.0.0, doctestplus-0.8.0, cov-2.11.1, hypothesis-6.8.1
collected 4 items / 3 deselected / 1 selected                                                                                    

control/freqplot.py F                                                                                                      [100%]

============================================================ FAILURES ============================================================
________________________________________ [doctest] control.freqplot.singular_values_plot _________________________________________
1068     Other Parameters
1069     ----------------
1070     grid : bool
1071         If True, plot grid lines on gain and phase plots.  Default is set by
1072         `config.defaults['freqplot.grid']`.
1073 
1074     Examples
1075     --------
1076     >>> den = [75, 1]
1077     >>> sys = ct.tf([[[87.8], [-86.4]], [[108.2], [-109.6]]], [[den, den], [den, den]])
UNEXPECTED EXCEPTION: NameError("name 'ct' is not defined")
Traceback (most recent call last):
  File "/usr/lib64/python3.8/doctest.py", line 1336, in __run
    exec(compile(example.source, filename, "single",
  File "<doctest control.freqplot.singular_values_plot[1]>", line 1, in <module>
NameError: name 'ct' is not defined
/home/ben/src/python-control/control/freqplot.py:1077: UnexpectedException
==================================================== short test summary info =====================================================
FAILED control/freqplot.py::control.freqplot.singular_values_plot
```

Reply:
OK I see. The new version should be doctest compliant. Now

pytest --doctest-modules -k singular_values_plot control/freqplot.py 

does not return any errors